src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMAlgorithm.java

```diff
@@ -0,0 +1,497 @@
+ *     via calling {@link CoverageModelEMAlgorithm {@link #runExpectation()}}
+
+ * (see CNV-methods.pdf for technical details).
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+
+public final class CoverageModelEMAlgorithm<S extends AlleleMetadataProducer & CallStringProducer & ScalarProducer> {
+
+    protected final Logger logger = LogManager.getLogger(CoverageModelEMAlgorithm.class);
+    protected final CoverageModelEMParams params;
+    protected final CoverageModelEMWorkspace<S> workspace;
+    protected EMAlgorithmStatus status;
+
+    /**
+     * Public constructor.
+     *
+     * @param params specification of the hyper-parameters and other details of the algorithm
+     * @param workspace workspace for actual calculations
+     */
+    public CoverageModelEMAlgorithm(@Nonnull final CoverageModelEMParams params,
+                                    @Nonnull CoverageModelEMWorkspace<S> workspace) {
+        this.params = Utils.nonNull(params, "Target coverage EM algorithm parameters must be non-null");
+        params.validate();
+        this.workspace = Utils.nonNull(workspace, "The coverage modeller EM algorithm workspace must be non-null");
+        this.status = EMAlgorithmStatus.TBD;
+        logger.info("EM algorithm initialized.");
+    }
+
+    /**
+     * Log the header for iteration info
+     */
+    private void showIterationHeader() {
+        final String header = String.format("%-15s%-20s%-20s%-20s%-20s", "Iterations", "Type", "Log Likelihood", "Update Size", "Misc.");
+        logger.info(header);
+        logger.info(StringUtils.repeat("=", header.length()));
+    }
+
+    /**
+     * Log the iteration info
+     *
+     * @param iter iteration number
+     * @param title iteration title
+     * @param logLikelihood current log likelihood
+     * @param updateSize current update size
+     * @param misc other miscellaneous info
+     */
+    private void showIterationInfo(final int iter, final String title, final double logLikelihood,
+                                   final double updateSize, final String misc) {
+        final String header = String.format("%-15d%-20s%-20.6e%-20.6e%-20s", iter, title, logLikelihood, updateSize, misc);
+        logger.info(header);
+    }
+
+    /**
+     * A generic runner of a step in the EM algorithm
+     *
+     * @param routine a routine to run
+     * @param miscFactory a function from the routine signal to a miscellaneous iteration info string
+     * @param title title of the routine
+     * @param iterInfo the iteration info class (will be updated)
+     */
+    private void runRoutine(@Nonnull final Supplier<SubroutineSignal> routine,
+                            @Nonnull final Function<SubroutineSignal, String> miscFactory,
+                            @Nonnull final String title,
+                            @Nonnull EMAlgorithmIterationInfo iterInfo) {
+        final SubroutineSignal sig = routine.get();
```

Comment:
Well, it is not a purely functional code. The data is _stored_ inside the workspace (hence, the name) and the compute blocks. The returned `SubroutineSignal` merely provides information about the update (number of iterations, update size, etc).